File: julia.py
# julia_starfish.py
# Python 3.13+
# Usage:
#   python julia_starfish.py [N] [M] [out.png]
# Defaults: N=1024, M=1024, out=julia_starfish.png
#
# Dependency:
#   pip install pillow

import math
import sys
from PIL import Image

def hsv_to_rgb(h: float, s: float, v: float) -> tuple[int, int, int]:
	# h in [0, 1)
	i = int(h * 6.0)
	f = h * 6.0 - i
	p = v * (1.0 - s)
	q = v * (1.0 - f * s)
	t = v * (1.0 - (1.0 - f) * s)

	i %= 6
	if i == 0:
		r, g, b = v, t, p
	elif i == 1:
		r, g, b = q, v, p
	elif i == 2:
		r, g, b = p, v, t
	elif i == 3:
		r, g, b = p, q, v
	elif i == 4:
		r, g, b = t, p, v
	else:
		r, g, b = v, p, q

	return (int(r * 255.0) & 255, int(g * 255.0) & 255, int(b * 255.0) & 255)

def compute_julia_starfish(n: int, m: int, out_path: str) -> None:
	# "Starfish" Julia constant (tunable)
	cr = -0.4
	ci =  0.6

	width = n
	height = n

	# View window (tunable)
	xmin, xmax = -1.6, 1.6
	ymin, ymax = -1.6, 1.6

	dx = (xmax - xmin) / (width - 1)
	dy = (ymax - ymin) / (height - 1)

	# Precompute coords
	xs = [xmin + x * dx for x in range(width)]
	ys = [ymax - y * dy for y in range(height)]  # flip Y

	img = Image.new("RGBA", (width, height))
	pix = img.load()

	escape2 = 4.0
	inv_log2 = 1.0 / math.log(2.0)

	for y in range(height):
		y0 = ys[y]
		for x in range(width):
			zr = xs[x]
			zi = y0

			i = 0
			zr2 = zr * zr
			zi2 = zi * zi

			while i < m and (zr2 + zi2) <= escape2:
				# z = z^2 + c
				zi = (zr + zr) * zi + ci
				zr = (zr2 - zi2) + cr

				zr2 = zr * zr
				zi2 = zi * zi
				i += 1

			if i >= m:
				pix[x, y] = (0, 0, 0, 255)
			else:
				mag = math.sqrt(zr2 + zi2)
				nu = i + 1.0 - math.log(math.log(mag)) * inv_log2
				t = nu / m
				if t < 0.0:
					t = 0.0
				elif t > 1.0:
					t = 1.0

				h = (0.66 + 1.4 * t) % 1.0
				s = 0.85
				v = 0.15 + 0.95 * t
				r, g, b = hsv_to_rgb(h, s, v)

				pix[x, y] = (r, g, b, 255)

	img.save(out_path, "PNG")

def main() -> int:
	try:
		n = int(sys.argv[1]) if len(sys.argv) > 1 else 1024
		m = int(sys.argv[2]) if len(sys.argv) > 2 else 1024
		out_path = sys.argv[3] if len(sys.argv) > 3 else "julia_starfish-python.png"
	except Exception:
		print("Usage: python julia_starfish.py [N] [M] [out.png]")
		return 2

	if n <= 0 or m <= 0:
		print("Invalid N or M.")
		return 2

	compute_julia_starfish(n, m, out_path)
	print(f"Wrote {out_path} ({n}x{n}, M={m})")
	return 0

if __name__ == "__main__":
	raise SystemExit(main())

File: julia.mjs
// julia_starfish.js
// Node.js 25+
// Usage:
//   node julia_starfish.js [N] [M] [out.png]
// Defaults: N=1024, M=1024, out=julia_starfish.png
//
// Install dependency:
//   npm i

import sharp from "sharp";

function hsvToRgb(h, s, v)
{
	// h in [0,1)
	const i = Math.floor(h * 6);
	const f = h * 6 - i;
	const p = v * (1 - s);
	const q = v * (1 - f * s);
	const t = v * (1 - (1 - f) * s);

	let r, g, b;
	switch (i % 6)
	{
		case 0: r = v; g = t; b = p; break;
		case 1: r = q; g = v; b = p; break;
		case 2: r = p; g = v; b = t; break;
		case 3: r = p; g = q; b = v; break;
		case 4: r = t; g = p; b = v; break;
		case 5: r = v; g = p; b = q; break;
	}
	return [
		Math.max(0, Math.min(255, (r * 255) | 0)),
		Math.max(0, Math.min(255, (g * 255) | 0)),
		Math.max(0, Math.min(255, (b * 255) | 0)),
	];
}

function computeJuliaStarfish(N, M, outPath)
{
	// "Starfish" Julia constant (tunable)
	const cr = -0.4;
	const ci =  0.6;

	const width = N;
	const height = N;

	// View window (tunable)
	const xmin = -1.6, xmax = 1.6;
	const ymin = -1.6, ymax = 1.6;

	const dx = (xmax - xmin) / (width - 1);
	const dy = (ymax - ymin) / (height - 1);

	const data = Buffer.allocUnsafe(width * height * 4);

	// Precompute coordinate arrays for less work in the inner loop
	const xs = new Float64Array(width);
	const ys = new Float64Array(height);

	for (let x = 0; x < width; x++) xs[x] = xmin + x * dx;
	for (let y = 0; y < height; y++) ys[y] = ymax - y * dy; // flip Y for conventional orientation

	const escape2 = 4.0;
	const invLog2 = 1.0 / Math.log(2);

	let p = 0;
	for (let y = 0; y < height; y++)
	{
		const y0 = ys[y];
		for (let x = 0; x < width; x++)
		{
			let zr = xs[x];
			let zi = y0;

			let i = 0;
			let zr2 = zr * zr;
			let zi2 = zi * zi;

			while (i < M && (zr2 + zi2) <= escape2)
			{
				// z = z^2 + c
				zi = (zr + zr) * zi + ci;
				zr = (zr2 - zi2) + cr;

				zr2 = zr * zr;
				zi2 = zi * zi;
				i++;
			}

			if (i >= M)
			{
				// inside
				data[p++] = 0;
				data[p++] = 0;
				data[p++] = 0;
				data[p++] = 255;
			}
			else
			{
				// Smooth coloring
				const mag = Math.sqrt(zr2 + zi2);
				const nu = i + 1 - Math.log(Math.log(mag)) * invLog2;
				const t = Math.max(0, Math.min(1, nu / M));

				// Hue ramp
				const h = (0.66 + 1.4 * t) % 1.0;
				const s = 0.85;
				const v = 0.15 + 0.95 * t;

				const [r, g, b] = hsvToRgb(h, s, v);

				data[p++] = r;
				data[p++] = g;
				data[p++] = b;
				data[p++] = 255;
			}
		}
	}

	return sharp(data, { raw: { width, height, channels: 4 } })
		.png()
		.toFile(outPath);
}

const N = Number(process.argv[2] ?? 1024);
const M = Number(process.argv[3] ?? 1024);
const outPath = String(process.argv[4] ?? "julia_starfish-node.png");

if (!Number.isFinite(N) || !Number.isFinite(M) || N <= 0 || M <= 0)
{
	console.error("Invalid N or M.");
	process.exit(1);
}

await computeJuliaStarfish(N | 0, M | 0, outPath);
console.log(`Wrote ${outPath} (${N}x${N}, M=${M})`);
